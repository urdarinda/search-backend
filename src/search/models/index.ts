import { Post, SearchInterface } from '../interface';
import { PostDatabase } from './schema';

const invertedIndex: Map<string, Set<number>> = new Map();

const searchPost: SearchInterface['searchPosts'] = (params) => {
  const { numberOfItems = 10, page = 0, query = undefined, sortKey } = params;

  return items;
};

const buildIndex = () => {
  const posts = PostDatabase.get();
  posts.forEach((post) => addPostToIndex(post));
  console.log(invertedIndex)
};

const addPostToIndex = (post: Post) => {
  const getTokensFromString = (s: string) => s.toLowerCase().trim().split(' ');
  const titleTokens = getTokensFromString(post.name);
  const descriptionTokens = getTokensFromString(post.description);

  const addTokensToIndex = (tokens: string[]) => {
    tokens.forEach((token) => {
      const documentIndexes = invertedIndex.get(token) ?? new Set<number>();
      documentIndexes.add(post.id);
      invertedIndex.set(token, documentIndexes);
    });
  };
  addTokensToIndex(titleTokens);
  addTokensToIndex(descriptionTokens);
};

export {
    buildIndex
}