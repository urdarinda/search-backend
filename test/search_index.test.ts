import { buildIndex } from '../src/search/models/build';
import { PostDatabase } from '../src/search/models/schema';
import * as Mocks from './schema.mock';

beforeEach(() => {
  jest.resetAllMocks();
  jest.spyOn(PostDatabase, 'getPostByIds').mockImplementation((ids) => {
    if (!ids) return Mocks.mockPosts;
    return Mocks.mockPosts.filter((post) => ids.some((i) => i == post.id));
  });
});

describe('correct builds search index from existing posts', () => {
  it('given a set of posts in a databse correctly builds index for unigram tokens', () => {
    expect(buildIndex()).toMatchSnapshot();
  });
});
