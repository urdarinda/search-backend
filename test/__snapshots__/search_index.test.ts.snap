// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correct builds search index from existing posts given a set of posts in a databse correctly builds index for unigram tokens 1`] = `
Map {
  "customer" => Set {
    1,
  },
  "assurance" => Set {
    1,
  },
  "liaison" => Set {
    1,
  },
  "my" => Set {
    1,
  },
  "lion" => Set {
    1,
    2,
  },
  "king" => Set {
    1,
    3,
  },
  "is" => Set {
    1,
    3,
  },
  "nnice" => Set {
    1,
  },
  "lord" => Set {
    2,
  },
  "oh" => Set {
    2,
  },
  "movie" => Set {
    2,
    3,
  },
  "of" => Set {
    2,
  },
  "the" => Set {
    2,
  },
  "rings" => Set {
    2,
  },
  "something" => Set {
    2,
  },
  "this" => Set {
    3,
  },
  "some" => Set {
    3,
  },
  "title" => Set {
    3,
  },
  "with" => Set {
    3,
  },
  "and" => Set {
    3,
  },
  "queen" => Set {
    3,
  },
  "both" => Set {
    3,
  },
  "are" => Set {
    3,
  },
  "awesome" => Set {
    3,
  },
}
`;
